File: 송하연/10월 4주차/거리두기 확인하기.java
import java.util.*;

class Solution {
    
    public int[] solution(String[][] places) {
        int[] answer = new int[places.length];
        
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};
        
        int idx=0;
        
        for(String[] place:places){
            char[][] map = new char[5][5];
            List<int[]> peopleList = new ArrayList<>(); 
            
            for(int i=0; i<place.length; i++){
                for(int j=0; j<place[i].length();j++){
                    map[i][j] = place[i].charAt(j);
                    if(map[i][j]=='P'){
                        peopleList.add(new int[]{i,j});
                    }
                }
            }
            
            int result = 1;
            
            for (int[] p : peopleList) {
                int x = p[0];
                int y = p[1];
                //System.out.println(x + "," + y);
                //System.out.println(bfs(x,y,dx,dy,map));
                if(!bfs(x,y,dx,dy,map)) {
                    result = 0;
                    break;
                }
            }
            answer[idx] =result;
            idx++;
        }
        
        return answer;
    }
    
    boolean bfs(int x, int y, int[] dx, int[] dy, char[][] map) {
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{x,y,0});
        
        boolean[][] visited = new boolean[5][5];
        visited[x][y] = true; 
        
        while(!queue.isEmpty()){
            int[] current = queue.poll();
            int cx = current[0];
            int cy = current[1];
            int d = current[2];
            
            for(int i=0; i<4; i++){
                int nx = cx+dx[i];
                int ny = cy+dy[i];
                int nd = d+1;
                
                if(nx>=0 && ny>=0 && nx<5 && ny<5 && nd<=2 &&(!visited[nx][ny])){
                    visited[nx][ny] = true;
                    
                    if(map[nx][ny]=='X'){
                        continue;
                    }
                    
                    if(map[nx][ny]=='P'){
                        return false;
                    }
                    
                    if(map[nx][ny]=='O'){
                        queue.offer(new int[]{nx, ny, nd});
                    }
                }
            }  
        }

        return true;
        
    }
}

File: 한승우/10월 4주차/거리두기 확인하기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

/*
각 대기실별로 거리두기를 지키고 있는지 확인한다

BFS 사용, 파티션 나오면 종료, 다른 사람에 도달하면 종료
*/

int visited[5][5];
int dx[] = {1, -1, 0, 0};
int dy[] = {0, 0, 1, -1};

bool in_range(int x, int y){
    return 0 <= x && x < 5 && 0 <= y && y < 5;
}

bool bfs(int r, int c, const vector<string>& place) {
    queue<pair<int, int>> q;
    
    //초기화
    for(int i = 0; i < 5; i++)
        for(int j = 0; j < 5; j++)
            visited[i][j] = 0;
    
    visited[r][c] = 1;
    q.push({r, c});
    
    while(!q.empty()){
        int x = q.front().first;
        int y = q.front().second;
        q.pop();
        
        //시작점이 아닌데 사람이라면
        if(visited[x][y] > 1 && place[x][y] == 'P') return false;
        //맨해튼 거리 2까지만 탐색
        if(visited[x][y] == 3) continue;
        
        for(int i = 0; i < 4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if(in_range(nx, ny) && !visited[nx][ny] && place[nx][ny] != 'X'){
                visited[nx][ny] = visited[x][y] + 1;
                q.push({nx, ny});
            }
        }
    }
    
    return true;
}

//한 대기실 탐색
bool room(const vector<string>& place){
    for(int i = 0; i < 5; i++){
        for(int j = 0; j < 5; j++){
            if(place[i][j] == 'P'){
                if(!bfs(i, j, place)){
                    return false;
                }
            }
        }
    }
    
    return true;
}

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    
    for(int k = 0; k < 5; k++){
        if(room(places[k])){
            answer.push_back(1);
        } else {
            answer.push_back(0);
        }
    }
    
    return answer; 
}